src/rollup/index.ts

```diff
@@ -174,184 +174,170 @@ function assignChunksToBundle(
 	return outputBundle as OutputBundle;
 }
 
-export default function rollup(rawInputOptions: GenericConfigObject): Promise<RollupBuild> {
-	try {
-		const inputOptions = getInputOptions(rawInputOptions);
-		initialiseTimers(inputOptions);
-
-		const graph = new Graph(inputOptions, curWatcher);
-		curWatcher = undefined as any;
-
-		// remove the cache option from the memory after graph creation (cache is not used anymore)
-		const useCache = rawInputOptions.cache !== false;
-		delete inputOptions.cache;
-		delete rawInputOptions.cache;
-
-		timeStart('BUILD', 1);
-
-		return graph.pluginDriver
-			.hookParallel('buildStart', [inputOptions])
-			.then(() =>
-				graph.build(
-					inputOptions.input as string | string[] | Record<string, string>,
-					inputOptions.manualChunks,
-					inputOptions.inlineDynamicImports as boolean
-				)
-			)
-			.then(
-				chunks => graph.pluginDriver.hookParallel('buildEnd', []).then(() => chunks),
-				err =>
-					graph.pluginDriver.hookParallel('buildEnd', [err]).then(() => {
-						throw err;
-					})
+export default async function rollup(rawInputOptions: GenericConfigObject): Promise<RollupBuild> {
+	const inputOptions = getInputOptions(rawInputOptions);
+	initialiseTimers(inputOptions);
+
+	const graph = new Graph(inputOptions, curWatcher);
+	curWatcher = undefined as any;
+
+	// remove the cache option from the memory after graph creation (cache is not used anymore)
+	const useCache = rawInputOptions.cache !== false;
+	delete inputOptions.cache;
+	delete rawInputOptions.cache;
+
+	timeStart('BUILD', 1);
+
+	const chunks = await graph.pluginDriver
+		.hookParallel('buildStart', [inputOptions])
+		.then(() =>
+			graph.build(
+				inputOptions.input as string | string[] | Record<string, string>,
+				inputOptions.manualChunks,
+				inputOptions.inlineDynamicImports as boolean
 			)
-			.then(chunks => {
-				timeEnd('BUILD', 1);
-
-				// ensure we only do one optimization pass per build
-				let optimized = false;
-
-				function getOutputOptions(rawOutputOptions: GenericConfigObject) {
-					return normalizeOutputOptions(
-						inputOptions as GenericConfigObject,
-						rawOutputOptions,
-						chunks.length > 1,
-						graph.pluginDriver
-					);
-				}
+		)
+		.then(
```

Comment:
Moving to a more async-await based flow is definitely something we want but this is not an improvement now that the different styles are mixed and we have async functions inside Promise chains inside an async function. If you complete the refactoring to use a `try-catch` block to catch the error and get rid of the remaining Promise chains in this section, I could see this merged.

Reply:
@lukastaegert You want this?

```typescript
	let chunks: Chunk[]

	try {
		await graph.pluginDriver.hookParallel('buildStart', [inputOptions]);
		chunks = await graph.build(
			inputOptions.input as string | string[] | Record<string, string>,
			inputOptions.manualChunks,
			inputOptions.inlineDynamicImports as boolean
		);
	} catch (err) {
		await graph.pluginDriver.hookParallel('buildEnd', [err]);
		throw err;
	}

	await graph.pluginDriver.hookParallel('buildEnd', []);
```

For reference, this is the original:

```typescript
	const chunks = await graph.pluginDriver
		.hookParallel('buildStart', [inputOptions])
		.then(() =>
			graph.build(
				inputOptions.input as string | string[] | Record<string, string>,
				inputOptions.manualChunks,
				inputOptions.inlineDynamicImports as boolean
			)
		)
		.then(
			async chunks => {
				await graph.pluginDriver.hookParallel('buildEnd', []);
				return chunks;
			},
			async err => {
				await graph.pluginDriver.hookParallel('buildEnd', [err]);
				throw err;
			}
		);
```

Reply:
Exactly 👍

Reply:
Well, I already done that (https://github.com/rollup/rollup/pull/3053/commits/4fcacdfc8fa4d57120278190cecd78aa96cc0bee).